src/ngx_http_lua_pipe.c

```diff
@@ -0,0 +1,2458 @@
+    }
+
+    return NGX_OK;
+}
+
+
+static void
+ngx_http_lua_pipe_proc_finalize(ngx_http_lua_ffi_pipe_proc_t *proc, int forced)
+{
+    ngx_http_lua_pipe_t          *pipe;
+
+    ngx_log_debug3(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
+                   "lua pipe finalize process:%p pid:%P forced:%d", proc,
+                   proc->_pid, forced);
+    pipe = proc->pipe;
+
+    if (pipe->node) {
+        ngx_rbtree_delete(&ngx_http_lua_pipe_rbtree, pipe->node);
+        pipe->node = NULL;
+    }
+
+    pipe->dead = 1;
+
+    ngx_http_lua_pipe_close_stdin(pipe, forced);
+    ngx_http_lua_pipe_close_stdout(pipe, forced);
+
+    if (!pipe->merge_stderr) {
+        ngx_http_lua_pipe_close_stderr(pipe, forced);
+    }
+
+    pipe->closed = 1;
+}
+
+
+void
+ngx_http_lua_pipe_proc_destroy(ngx_http_lua_ffi_pipe_proc_t *proc)
+{
+    ngx_http_lua_pipe_t          *pipe;
+
+    pipe = proc->pipe;
+    if (pipe == NULL) {
+        return;
+    }
+
+    ngx_log_debug2(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
+                   "lua pipe destroy process:%p pid:%P", proc, proc->_pid);
+
+    if (!pipe->dead) {
+        ngx_log_debug2(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
+                       "lua pipe kill process:%p pid:%P", proc, proc->_pid);
+
+        if (kill(proc->_pid, SIGKILL) == -1) {
+            if (ngx_errno != ESRCH) {
+                ngx_log_error(NGX_LOG_ERR, ngx_cycle->log, ngx_errno,
+                              "lua pipe failed to kill process:%p pid:%P");
+            }
+        }
+    }
+
+    ngx_http_lua_pipe_proc_finalize(proc, 1);
+    ngx_destroy_pool(pipe->pool);
+    proc->pipe = NULL;
+}
+
+
+static ngx_int_t
+ngx_http_lua_pipe_get_lua_ctx(ngx_http_request_t *r,
+    ngx_http_lua_ctx_t **ctx, u_char *errbuf, size_t *errbuf_size)
+{
+    int                                 rc;
+
+    *ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        return NGX_HTTP_LUA_FFI_NO_REQ_CTX;
+    }
+
+    rc = ngx_http_lua_ffi_check_context(*ctx, NGX_HTTP_LUA_CONTEXT_REWRITE
+                                        | NGX_HTTP_LUA_CONTEXT_ACCESS
+                                        | NGX_HTTP_LUA_CONTEXT_CONTENT
+                                        | NGX_HTTP_LUA_CONTEXT_TIMER
+                                        | NGX_HTTP_LUA_CONTEXT_SSL_CERT
+                                        | NGX_HTTP_LUA_CONTEXT_SSL_SESS_FETCH,
+                                        errbuf, errbuf_size);
+    if (rc != NGX_OK) {
+        return NGX_HTTP_LUA_FFI_BAD_CONTEXT;
+    }
+
+    return NGX_OK;
+}
+
+
+static void
+ngx_http_pipe_put_error(ngx_http_lua_pipe_ctx_t *pipe_ctx, u_char *errbuf,
+    size_t *errbuf_size)
+{
+    switch (pipe_ctx->err_type) {
+
+    case NGX_HTTP_LUA_PIPE_ET_CLOSED:
+        *errbuf_size =
+            ngx_snprintf(errbuf, *errbuf_size, "closed") - errbuf;
+        break;
+
+    case NGX_HTTP_LUA_PIPE_ET_ERROR:
+        *errbuf_size = ngx_snprintf(errbuf, *errbuf_size, "error") - errbuf;
```

Comment:
@thibaultcha 
Well, errors which are rare and could not be classified will be marked as `NGX_HTTP_LUA_PIPE_ET_ERROR`. Because we always log the error when this type of error happened, it is acceptable to use `"error"` as the returned error message.

Reply:
@spacewander Can we avoid automatic logging but return the detailed error message to the Lua caller? It's much easier to handle errors on the Lua land than messing up with error log messages. The cosocket API also provides the `lua_socket_log_errors` directive to turn off automatic logging completely.